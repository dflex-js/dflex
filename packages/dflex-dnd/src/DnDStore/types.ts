import type { Dimensions, ITracker } from "@dflex/utils";
import type { IDFlexContainer, IDFlexNode } from "@dflex/core-instance";

import type { RegisterInputOpts, IDFlexBaseStore } from "@dflex/store";
import type { DFlexListenersInitializer } from "./DFlexListeners";

export type ELmKey = string;

export interface ElmTree {
  element: IDFlexNode;
  parent: IDFlexNode | null;
  branches: {
    siblings: string[];
    parents: string[];
  };
}

export interface ScrollThreshold {
  maxX: number;
  maxY: number;
  minX: number;
  minY: number;
}

interface Translate {
  translateX: number;
  translateY: number;
}

export interface IDFlexDnDStore extends IDFlexBaseStore {
  readonly containers: Map<string, IDFlexContainer>;
  readonly unifiedContainerDimensions: Map<number, Dimensions>;
  readonly tracker: ITracker;
  observer: MutationObserver | null;
  listeners: DFlexListenersInitializer;
  initSiblingContainer(SK: string, shouldValidate: boolean): void;
  updateBranchVisibility(SK: string, shouldCheckVisibility: boolean): void;
  register(element: RegisterInputOpts): void;
  getELmTranslateById(id: string): Translate;
  getElmTreeById(id: string): ElmTree;
  getElmSiblingsById(id: string): string[] | null;
  destroy(): void;
}
