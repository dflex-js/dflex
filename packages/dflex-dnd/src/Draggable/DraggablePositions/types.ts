export type MovementDirection = "r" | "l" | "d" | "u";
