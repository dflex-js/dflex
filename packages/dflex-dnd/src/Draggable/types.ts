import type {
  IPointNum,
  ThresholdInterface,
  ThresholdCoordinate,
  IMigration,
} from "@dflex/utils";
import type { IDFlexNode } from "@dflex/core-instance";
import type { IDFlexBaseDraggable } from "@dflex/draggable";

import type { ContainersTransition, ScrollOpts } from "../types";
import { dispatchDFlexEvent } from "../DnDStore";

export interface SiblingsThreshold {
  [sk: string]: ThresholdCoordinate;
}

export interface IDraggableAxes extends IDFlexBaseDraggable<IDFlexNode> {
  /** Dragged threshold  */
  readonly threshold: ThresholdInterface;

  readonly innerOffset: IPointNum;

  /** Dragged position for both X and Y.  */
  readonly positionPlaceholder: IPointNum;

  readonly migration: IMigration;

  /** grid placeholder for dragged grid position. */
  readonly gridPlaceholder: IPointNum;

  /** Restrict dragged movement inside viewport.  */
  readonly isViewportRestricted: boolean;

  readonly events: ReturnType<typeof dispatchDFlexEvent>;

  dragAt(x: number, y: number): void;

  /**
   * Check if the dragged out self position or parent container and set the
   * necessary flags.
   */
  isOutThreshold(SK?: string): boolean;
  isOutThreshold(SK: string, useInsertionThreshold: true): boolean;

  /** Has moved without settling inside new position. */
  isNotSettled(): boolean;
}

export interface IDraggableInteractive extends IDraggableAxes {
  readonly containersTransition: ContainersTransition;
  readonly scroll: ScrollOpts;
  readonly occupiedPosition: IPointNum;
  readonly occupiedTranslate: IPointNum;
  readonly isDraggedPositionFixed: boolean;
  setDraggedTempIndex(i: number): void;
  endDragging(isFallback: boolean): void;
}
