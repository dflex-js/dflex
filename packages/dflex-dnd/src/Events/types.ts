import { DFLEX_EVENTS, DFLEX_EVENTS_CAT, DFLEX_ATTRS } from "./constants";

const { DRAG_EVENT, INTERACTIVITY_EVENT, SIBLINGS_EVENT } = DFLEX_EVENTS;
const { DRAG_CAT, INTERACTIVITY_CAT, SIBLINGS_CAT } = DFLEX_EVENTS_CAT;
const { DRAG_ATTR } = DFLEX_ATTRS;

export type DragEventNames = (typeof DRAG_EVENT)[keyof typeof DRAG_EVENT];

export type InteractivityEventNames =
  (typeof INTERACTIVITY_EVENT)[keyof typeof INTERACTIVITY_EVENT];

export type SiblingsEventNames =
  (typeof SIBLINGS_EVENT)[keyof typeof SIBLINGS_EVENT];

export type DragAttr = (typeof DRAG_ATTR)[keyof typeof DRAG_ATTR];

/**
 * All available DFlex event names.
 */
export type DFlexEventNames =
  | DragEventNames
  | InteractivityEventNames
  | SiblingsEventNames;

export type PayloadDraggedEvent = {
  type: typeof DRAG_CAT;
  /** Returns element id in the registry */
  id: string;
  /** Returns dragged temp index */
  index: number;
};

export type PayloadInteractivityEvent = {
  type: typeof INTERACTIVITY_CAT;
  /** Returns element id in the registry */
  id: string;
  /** Returns element current index */
  index: number;
  /** Returns the element that triggered the event */
  target: HTMLElement;
};

export type PayloadSiblingsEvent = {
  type: typeof SIBLINGS_CAT;
  /** Returns the index where the dragged left */
  from: number;
  /** Returns the last index effected of the dragged leaving/entering */
  to: number;
  /** Returns an array of sibling ids in order */
  siblings: string[];
};

export type DFlexEventPayloads =
  | PayloadDraggedEvent
  | PayloadInteractivityEvent
  | PayloadSiblingsEvent;

/**
 * Event fired when drag exits threshold or container.
 * Contains drag event payload.
 */
export type DFlexDraggedEvent = CustomEvent<PayloadDraggedEvent>;

/**
 * Event fired when drag enters or leaves an element.
 * Contains interactivity event payload.
 */
export type DFlexInteractivityEvent = CustomEvent<PayloadInteractivityEvent>;

/**
 * Event fired when drag triggers sibling rearrangement.
 * Contains sibling event payload.
 */
export type DFlexSiblingsEvent = CustomEvent<PayloadSiblingsEvent>;

/**
 * Union of all DFlex drag, interactivity and sibling events.
 */
export type DFlexEvents = CustomEvent<
  PayloadDraggedEvent | PayloadInteractivityEvent | PayloadSiblingsEvent
>;

export type DFlexEventsMap = {
  [K in DragEventNames]: DFlexDraggedEvent;
} & {
  [K in InteractivityEventNames]: DFlexInteractivityEvent;
} & {
  [K in SiblingsEventNames]: DFlexSiblingsEvent;
};
