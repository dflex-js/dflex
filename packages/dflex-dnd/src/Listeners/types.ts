import { DFLEX_LISTENERS_CAT, LayoutStates, MutationStates } from "./constants";

export type LayoutState = (typeof LayoutStates)[keyof typeof LayoutStates];

export type MutationState =
  (typeof MutationStates)[keyof typeof MutationStates];

export type DFlexLayoutStateNotification = {
  type: typeof DFLEX_LISTENERS_CAT.LAYOUT_CAT;
  status: LayoutState;
  payload?: never;
};

export type DFlexElmMutationNotification = {
  type: typeof DFLEX_LISTENERS_CAT.MUTATION_CAT;
  status: MutationState;
  payload: {
    target: HTMLElement; // HTML element container.
    ids: string[]; // Committed Elements' id in order.
  };
};

export type DFlexErrorNotification = {
  type: typeof DFLEX_LISTENERS_CAT.ERROR_CAT;
  error: Error;
};

export type DFlexListenerNotifications =
  | DFlexLayoutStateNotification
  | DFlexElmMutationNotification
  | DFlexErrorNotification;
