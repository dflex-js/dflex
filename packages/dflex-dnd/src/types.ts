import type { ThresholdPercentages } from "@dflex/utils";
import type { DFlexEventsTypes } from "./LayoutManager";

type DFlexEventsMap = {
  [K in DFlexEventsTypes]: CustomEvent<any>;
};

declare global {
  interface Document {
    addEventListener<key extends keyof DFlexEventsMap>(
      DFlexEventType: key,
      listener: (this: Document, evt: DFlexEventsMap[key]) => void
    ): void;
    removeEventListener<K extends keyof DFlexEventsMap>(
      DFlexEventType: K,
      listener: (this: Document, evt: DFlexEventsMap[K]) => void
    ): void;
  }
}

interface RestrictionsMeta {
  readonly allowLeavingFromTop: boolean;
  readonly allowLeavingFromBottom: boolean;
  readonly allowLeavingFromLeft: boolean;
  readonly allowLeavingFromRight: boolean;
}

export interface Restrictions {
  self: RestrictionsMeta;
  container: RestrictionsMeta;
}

export interface RestrictionsPartial {
  self?: Partial<RestrictionsMeta>;
  container?: Partial<RestrictionsMeta>;
}

export interface ScrollOptsBase {
  enable: boolean;
  initialSpeed: number;
}

export interface ScrollOptPartialThreshold extends ScrollOptsBase {
  threshold: Partial<ThresholdPercentages>;
}

export interface ScrollOpts extends ScrollOptsBase {
  threshold: ThresholdPercentages;
}

export interface ContainersTransition {
  /** Default=true */
  enable: boolean;

  /**
   * Support orphan to orphan transformation.
   * Default=10px
   * */
  margin: number;
}

export interface DefaultDndOpts {
  containersTransition: ContainersTransition;
  threshold: ThresholdPercentages;
  restrictions: Restrictions;
  scroll: ScrollOpts;
}

export interface RestrictionsStatus {
  isContainerRestricted: boolean;
  isSelfRestricted: boolean;
}

export interface FinalDndOpts extends DefaultDndOpts {
  restrictionsStatus: RestrictionsStatus;
}

export interface DFlexDnDOpts {
  containersTransition?: Partial<ContainersTransition>;
  threshold?: Partial<ThresholdPercentages>;
  restrictions?: RestrictionsPartial;
  scroll?: Partial<ScrollOptPartialThreshold>;
}
