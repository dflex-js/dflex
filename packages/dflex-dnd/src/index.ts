export { exportedStoreSingleton as store } from "./store";

export { default as DnD } from "./DnD";

export type { DFlexDnDOpts } from "./types";

export type {
  DFlexDraggedEvent,
  DFlexInteractivityEvent,
  DFlexSiblingsEvent,
  DFlexEvents,
  DFlexEventsTypes,
  DFlexLayoutStateEvent,
  DFlexElmMutationEvent,
  DFlexListenerEvents,
} from "./LayoutManager";

export type {
  DFlexScrollContainer,
  DFlexSerializedElement,
  DFlexDOMGenOrder,
} from "@dflex/core-instance";

export type { RegisterInputOpts, DFlexGlobalConfig } from "@dflex/store";

export type {
  AxesPoint,
  AbstractBoxRect as BoxRectAbstract,
  PointNum,
  AbstractBox,
  Dimensions,
} from "@dflex/utils";
