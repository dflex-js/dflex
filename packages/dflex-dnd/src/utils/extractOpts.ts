// @ts-nocheck

// TODO: Add types for this.

import type { DndOpts, FinalDndOpts } from "../types";
import { defaultOpts, noop } from "./constants";

export function extractOpts(opts: DndOpts) {
  const options = { ...opts };

  (Object.keys(defaultOpts) as Array<keyof FinalDndOpts>).forEach((props) => {
    if (props === "restrictions") {
      options.restrictionsStatus = {
        isSelfRestricted: false,
        isContainerRestricted: false,
      };
    }

    if (options[props] === undefined) {
      options[props] =
        typeof options[props] === "object"
          ? { ...defaultOpts[props] }
          : defaultOpts[props];

      return;
    }

    Object.keys(defaultOpts[props]).forEach((subProp) => {
      // for sub options with objects values, like restrictions.
      if (typeof defaultOpts[props][subProp] === "object") {
        if (options[props][subProp] !== undefined) {
          options[props][subProp] = {
            ...defaultOpts[props][subProp],
            ...options[props][subProp],
          };

          return;
        }

        options[props][subProp] = { ...defaultOpts[props][subProp] };

        return;
      }

      if (options[props][subProp] === undefined) {
        options[props][subProp] = defaultOpts[props][subProp];
      }
    });

    if (props === "restrictions") {
      const restrictionsKeys = Object.keys(options.restrictions.self);

      let isSelfRestricted = false;

      for (let i = 0; i < restrictionsKeys.length; i += 1) {
        if (!options[props].self[restrictionsKeys[i]]) {
          isSelfRestricted = true;

          break;
        }
      }

      let isContainerRestricted = false;

      for (let i = 0; i < restrictionsKeys.length; i += 1) {
        if (!options[props].container[restrictionsKeys[i]]) {
          isContainerRestricted = true;

          break;
        }
      }

      options.restrictionsStatus.isSelfRestricted = isSelfRestricted;
      options.restrictionsStatus.isContainerRestricted = isContainerRestricted;
    } else if (props === "events") {
      Object.keys(options.events).forEach((event) => {
        if (typeof options.events[event] !== "function") {
          options.events[event] = noop;
        }
      });
    }
  });

  return options as FinalDndOpts;
}

export default extractOpts;
