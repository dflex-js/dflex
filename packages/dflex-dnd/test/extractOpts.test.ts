import { InteractivityEvent } from "../src/types";
import { extractOpts } from "../src/utils/extractOpts";

describe("extractOpts", () => {
  it("Returns default options when empty object is passed", () => {
    const opts = extractOpts({});

    expect(opts).toMatchSnapshot();
  });

  it("Returns the default options with input scroll/threshold", () => {
    const opts = extractOpts({ scroll: { threshold: { horizontal: 1000 } } });

    expect(opts).toMatchSnapshot();

    expect(opts.scroll.threshold.horizontal).toEqual(1000);
  });

  it("Returns the default options with input scroll/threshold/enable", () => {
    const opts = extractOpts({
      scroll: { enable: false, threshold: { horizontal: 1000 } },
    });

    expect(opts).toMatchSnapshot();

    expect(opts.scroll.threshold.horizontal).toEqual(1000);
    expect(opts.scroll.enable).toEqual(false);
  });

  it("Returns the default options with self restrictions/allowLeavingFromLeft+right", () => {
    const opts = extractOpts({
      restrictions: {
        self: {
          allowLeavingFromTop: true,
          allowLeavingFromBottom: true,
          allowLeavingFromLeft: false,
          allowLeavingFromRight: false,
        },
      },
    });

    expect(opts).toMatchSnapshot();

    expect(opts.restrictions.self.allowLeavingFromLeft).toEqual(false);
    expect(opts.restrictions.self.allowLeavingFromRight).toEqual(false);
  });

  it("Returns the default options with input restrictions/allowLeavingFromBottom", () => {
    const opts = extractOpts({
      restrictions: {
        container: {
          allowLeavingFromBottom: false,
        },
      },
    });

    expect(opts).toMatchSnapshot();

    expect(opts.restrictions.container.allowLeavingFromBottom).toEqual(false);
  });

  it("Returns the default options with input restrictions/threshold/scroll", () => {
    const opts = extractOpts({
      containersTransition: {
        enable: false,
      },
      restrictions: {
        container: {
          allowLeavingFromBottom: false,
        },
        self: {
          allowLeavingFromTop: false,
        },
      },
      threshold: {
        horizontal: 9000,
      },
      scroll: {
        threshold: {
          horizontal: 9000,
        },
      },
    });

    expect(opts).toMatchSnapshot();

    expect(opts.restrictions.container.allowLeavingFromBottom).toEqual(false);
    expect(opts.restrictions.self.allowLeavingFromTop).toEqual(false);

    expect(opts.containersTransition).toMatchInlineSnapshot(`
      Object {
        "enable": false,
        "margin": 10,
      }
    `);

    expect(opts.threshold.horizontal).toEqual(9000);
    expect(opts.scroll.threshold.horizontal).toEqual(9000);
  });

  it("All three events handlers are defined", () => {
    const opts = extractOpts({});

    expect(opts.events.onDragLeave).toBeDefined();
    expect(opts.events.onDragOver).toBeDefined();
    expect(opts.events.onStateChange).toBeDefined();
  });

  it("Returns the correct event handler function", () => {
    function onDragLeave(e: InteractivityEvent) {
      // eslint-disable-next-line no-console
      console.log(e);
    }

    const opts = extractOpts({
      events: {
        onDragLeave,
      },
    });

    expect(opts).toMatchSnapshot();
  });
});
