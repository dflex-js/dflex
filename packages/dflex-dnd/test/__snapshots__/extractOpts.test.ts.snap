// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractOpts Returns default options when empty object is passed 1`] = `
{
  "commit": {
    "enableAfterEndingDrag": true,
    "enableForScrollOnly": true,
  },
  "containersTransition": {
    "enable": true,
    "margin": 10,
  },
  "restrictions": {
    "container": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
    "self": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
  },
  "restrictionsStatus": {
    "isContainerRestricted": false,
    "isSelfRestricted": false,
  },
  "scroll": {
    "enable": true,
    "initialSpeed": 10,
    "threshold": {
      "horizontal": 15,
      "vertical": 15,
    },
  },
  "threshold": {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;

exports[`extractOpts Returns the default options with input restrictions/allowLeavingFromBottom 1`] = `
{
  "commit": {
    "enableAfterEndingDrag": true,
    "enableForScrollOnly": true,
  },
  "containersTransition": {
    "enable": true,
    "margin": 10,
  },
  "restrictions": {
    "container": {
      "allowLeavingFromBottom": false,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
    "self": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
  },
  "restrictionsStatus": {
    "isContainerRestricted": true,
    "isSelfRestricted": false,
  },
  "scroll": {
    "enable": true,
    "initialSpeed": 10,
    "threshold": {
      "horizontal": 15,
      "vertical": 15,
    },
  },
  "threshold": {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;

exports[`extractOpts Returns the default options with input restrictions/threshold/scroll 1`] = `
{
  "commit": {
    "enableAfterEndingDrag": true,
    "enableForScrollOnly": true,
  },
  "containersTransition": {
    "enable": false,
    "margin": 10,
  },
  "restrictions": {
    "container": {
      "allowLeavingFromBottom": false,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
    "self": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": false,
    },
  },
  "restrictionsStatus": {
    "isContainerRestricted": true,
    "isSelfRestricted": true,
  },
  "scroll": {
    "enable": true,
    "initialSpeed": 10,
    "threshold": {
      "horizontal": 9000,
      "vertical": 15,
    },
  },
  "threshold": {
    "horizontal": 9000,
    "vertical": 60,
  },
}
`;

exports[`extractOpts Returns the default options with input scroll/threshold 1`] = `
{
  "commit": {
    "enableAfterEndingDrag": true,
    "enableForScrollOnly": true,
  },
  "containersTransition": {
    "enable": true,
    "margin": 10,
  },
  "restrictions": {
    "container": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
    "self": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
  },
  "restrictionsStatus": {
    "isContainerRestricted": false,
    "isSelfRestricted": false,
  },
  "scroll": {
    "enable": true,
    "initialSpeed": 10,
    "threshold": {
      "horizontal": 1000,
      "vertical": 15,
    },
  },
  "threshold": {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;

exports[`extractOpts Returns the default options with input scroll/threshold/enable 1`] = `
{
  "commit": {
    "enableAfterEndingDrag": true,
    "enableForScrollOnly": true,
  },
  "containersTransition": {
    "enable": true,
    "margin": 10,
  },
  "restrictions": {
    "container": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
    "self": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
  },
  "restrictionsStatus": {
    "isContainerRestricted": false,
    "isSelfRestricted": false,
  },
  "scroll": {
    "enable": false,
    "initialSpeed": 10,
    "threshold": {
      "horizontal": 1000,
      "vertical": 15,
    },
  },
  "threshold": {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;

exports[`extractOpts Returns the default options with self restrictions/allowLeavingFromLeft+right 1`] = `
{
  "commit": {
    "enableAfterEndingDrag": true,
    "enableForScrollOnly": true,
  },
  "containersTransition": {
    "enable": true,
    "margin": 10,
  },
  "restrictions": {
    "container": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": true,
      "allowLeavingFromRight": true,
      "allowLeavingFromTop": true,
    },
    "self": {
      "allowLeavingFromBottom": true,
      "allowLeavingFromLeft": false,
      "allowLeavingFromRight": false,
      "allowLeavingFromTop": true,
    },
  },
  "restrictionsStatus": {
    "isContainerRestricted": false,
    "isSelfRestricted": true,
  },
  "scroll": {
    "enable": true,
    "initialSpeed": 10,
    "threshold": {
      "horizontal": 15,
      "vertical": 15,
    },
  },
  "threshold": {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;
